test/mpi/coll/allred.c

```diff
@@ -40,367 +40,549 @@ struct double_test {
+            err += cmp_complex(p, q, type_size);
+        } else if (category == CATEGORY_PAIR) {
+            err += cmp_pair(p, q, mpi_type);
+        } else {
+            assert(0);
+        }
+        p += type_size;
+        q += type_size;
     }
+    return err;
+}
 
-#if MTEST_HAVE_MIN_MPI_VERSION(2,2)
-#define test_types_set_mpi_2_2_integer(op,post) do {    \
-        op##_test##post(int8_t, MPI_INT8_T);            \
-        op##_test##post(int16_t, MPI_INT16_T);          \
-        op##_test##post(int32_t, MPI_INT32_T);          \
-        op##_test##post(int64_t, MPI_INT64_T);          \
-        op##_test##post(uint8_t, MPI_UINT8_T);          \
-        op##_test##post(uint16_t, MPI_UINT16_T);        \
-        op##_test##post(uint32_t, MPI_UINT32_T);        \
-        op##_test##post(uint64_t, MPI_UINT64_T);        \
-        op##_test##post(MPI_Aint, MPI_AINT);            \
-        op##_test##post(MPI_Offset, MPI_OFFSET);        \
-    } while (0)
-#else
-#define test_types_set_mpi_2_2_integer(op,post) do { } while (0)
-#endif
+static void report_error(int rank, MPI_Datatype mpi_type, MPI_Op op)
+{
+    char name[MPI_MAX_OBJECT_NAME] = { 0 };
+    int len = 0;
 
-#if MTEST_HAVE_MIN_MPI_VERSION(3,0)
-#define test_types_set_mpi_3_0_integer(op,post) do {    \
-        op##_test##post(MPI_Count, MPI_COUNT);          \
-    } while (0)
-#else
-#define test_types_set_mpi_3_0_integer(op,post) do { } while (0)
-#endif
+    MPI_Type_get_name(mpi_type, name, &len);
+    printf("(%d) Error for type %s and op %s\n", rank, name, mpi_op2str(op));
+}
 
-#define test_types_set1(op, post)                               \
-    {                                                           \
-        op##_test##post(int, MPI_INT);                          \
-        op##_test##post(long, MPI_LONG);                        \
-        op##_test##post(short, MPI_SHORT);                      \
-        op##_test##post(unsigned short, MPI_UNSIGNED_SHORT);    \
-        op##_test##post(unsigned, MPI_UNSIGNED);                \
-        op##_test##post(unsigned long, MPI_UNSIGNED_LONG);      \
-        op##_test##post(unsigned char, MPI_UNSIGNED_CHAR);      \
-        test_types_set_mpi_2_2_integer(op,post);                \
-        test_types_set_mpi_3_0_integer(op,post);                \
-    }
+/* The logic on the error check on MPI_Allreduce assumes that all
+   MPI_Allreduce routines return a failure if any do - this is sufficient
+   for MPI implementations that reject some of the valid op/datatype pairs
+   (and motivated this addition, as some versions of the IBM MPI
+   failed in just this way).
+*/
 
-#define test_types_set2(op, post)               \
-    {                                           \
-        test_types_set1(op, post);              \
-        op##_test##post(float, MPI_FLOAT);      \
-        op##_test##post(double, MPI_DOUBLE);    \
+static int allreduce_and_check(MPI_Datatype mpi_type, int category, MPI_Op op, int memtype)
```

Comment:
this function doesn't look like it returns a value

Reply:
Will address together with the next comment.